x-pack/functionbeat/manager/gcp/cli_manager.go

```diff
@@ -0,0 +1,169 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package gcp
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/x-pack/functionbeat/function/provider"
+	"github.com/elastic/beats/x-pack/functionbeat/manager/executor"
+	fngcp "github.com/elastic/beats/x-pack/functionbeat/provider/gcp/gcp"
+)
+
+type installer interface {
+	Config() *fngcp.FunctionConfig
+	Name() string
+}
+
+// CLIManager interacts with the AWS Lambda API to deploy, update or remove a function.
+// It will take care of creating the main lambda function and ask for each function type for the
+// operation that need to be executed to connect the lambda to the triggers.
+type CLIManager struct {
+	templateBuilder *defaultTemplateBuilder
+	location        string
+	tokenSrc        oauth2.TokenSource
+	log             *logp.Logger
+	config          *Config
+}
+
+// Deploy delegates deploy to the actual function implementation.
+func (c *CLIManager) Deploy(name string) error {
+	c.log.Debugf("Deploying function: %s", name)
+	defer c.log.Debugf("Deploy finish for function '%s'", name)
+
+	create := false
+	err := c.deploy(create, name)
+	if err != nil {
+		return err
+	}
+
+	c.log.Debugf("Successfully created function '%s'", name)
+	return nil
+}
+
+// Update updates the function.
+func (c *CLIManager) Update(name string) error {
+	c.log.Debugf("Starting updating function '%s'", name)
+	defer c.log.Debugf("Update complete for function '%s'", name)
+
+	update := true
+	err := c.deploy(update, name)
+	if err != nil {
+		return err
+	}
+
+	c.log.Debugf("Successfully updated function: '%s'", name)
+	return nil
+}
+
+// deploy uploads to bucket and creates a function on GCP.
+func (c *CLIManager) deploy(update bool, name string) error {
+	functionData, err := c.templateBuilder.execute(name)
+	if err != nil {
+		return err
+	}
+
+	executer := executor.NewExecutor(c.log)
+	executer.Add(newOpEnsureBucket(c.log, c.config))
+	executer.Add(newOpUploadToBucket(c.log, c.config, name, functionData.raw))
+
+	if update {
+		executer.Add(newOpUpdateFunction(c.log, c.tokenSrc, functionData.function.Name, functionData.function))
+	} else {
+		executer.Add(newOpCreateFunction(c.log, c.tokenSrc, c.location, name, functionData.function))
+	}
+
+	executer.Add(newOpWaitForFunction(c.log, c.tokenSrc))
+
+	ctx := &functionContext{}
+	if err := executer.Execute(ctx); err != nil {
```

Comment:
Do we need to pass functionContext via Execute? The way operations are setup we could pass a reference to shared state/info to the operations constructors instead.

Reply:
You mean in `executor.NewExecutor`?

Reply:
No. One can also setup some shared context via `newOpX`. Given that operations setup + execution are still executed in the same context, I'd consider the `ctx` argument passed `Execute` is not really required.

Reply:
I see. Indeed, we could pass it via the constructors. What is the advantage of passing it that way? 3 out of 5 operations need context, so most of those would need adjustments. The whole executor runs for 5-10 minutes tops. We don't necessarily need to optimize for memory, performance, etc. Also, if we decide to use this approach, the manager of AWS should be adjusted as well, for the sake of consistency.

Reply:
It is not about memory or performance. It's more a matter of type safety (better have the compiler check a type, then postpone it to runtime if possible) + making things more explicitly visible in code. By passing shared state via context we have to type cast, plus we can't tell which operations use the shared context or not. Registering the shared state with the constructor makes it explicit where it comes from, which operations use it + there is no extra type-cast per op.

Via defensive programming (checking the type cast being correct) we even hide a potential programmer error behind an error value (errWrongContext) in the opX implementations. If this error is printed to console, can we tell where it comes from and how to resolve it easily? Sometimes it is better to be 'loud' about a programmer error, instead of hiding it behind error values that look like a normal errors (e.g. have a stack trace)?

Reply:
I am passing the context via the constructors. I am submitting a separate PR with the adjustments in AWS.